### ISSUE TITLE
Regularly occurring JMS error: startTransaction (XA) on JMSService:jmsdirect failed

### ISSUE THREAD

**Issue by dorpsidioot** [payara/Payara#380]
I'm encountering the following error(see logdump below) when sending JMS messages in bulk in order to trigger some action(as you would do with JMS). The glassfish 4.x issue tracker lists this issue as resolved(see https://java.net/jira/browse/GLASSFISH-19872), however I am still experiencing this. Mind informing me what is going on, new to EE7 and JMSContext so it is possible that I committed a brainfart here, but I'd like to be sure.

### Code which triggers the JMS messages

```
@Stateless
@LocalBean
public class AggregationNotifier {

    @EJB
    private AggregationController controller;

    @Resource(mappedName = "jms/b2bev/aggregation")
    private Queue aggregationQueue;

    @Inject
    private JMSContext context;

    private static final Logger log = Logger.getLogger(AggregationNotifier.class.getName());

    public void trigger(List<Integer> companyIdentifiers, DateTime startDate, DateTime endDate) {
        try {
            if (companyIdentifiers != null) {
                for (Integer companyId : companyIdentifiers) {
                    AggregationEvent event = new AggregationEvent(companyId, startDate, endDate);
                    // Serializing payload to JSON
                    String payload = SerializingHandler.serialize(event);
                    sendMessageFurther(payload, aggregationQueue);
                }
            }
        } catch (Exception ex) {
            log.log(Level.WARNING, "An error has occurred while sending notification - Error: {0}", ex.getMessage());
        }
    }

    private void sendMessageFurther(String payload, Queue targetQueue) {

        try {
            context.createProducer().send(targetQueue, payload);
        } catch (JMSRuntimeException ex) {
            log.log(Level.WARNING, "sendMessageFurther encountered an exception - Error: {0}", ex.getMessage());
        }
    }
}
```
### MDB which handles the messages

```
@MessageDriven(activationConfig = {
    @ActivationConfigProperty(propertyName = "destinationType", propertyValue = "javax.jms.Queue"),
    @ActivationConfigProperty(propertyName = "destinationLookup", propertyValue = "jms/b2bev/aggregation")
})
public class AggregationMessageListener implements MessageListener {

    @EJB
    private AggregationController aggregationController;

    private static final Logger log = Logger.getLogger(AggregationMessageListener.class.getName());

    @Override
    public void onMessage(Message message) {
        if (message instanceof TextMessage) {

            try {

                String messageText = message.getBody(String.class);
                log.log(Level.INFO, "Received message: {0}", new Object[]{messageText});
                Object raw = SerializingHandler.deserialize(messageText, AggregationEvent.class);
                if (raw != null && raw instanceof AggregationEvent) {
                    AggregationEvent event = (AggregationEvent) raw;
                    AggregationHandler handler = new AggregationHandler();

                    if (event.getStartDate() != null && event.getStartDate() != null) {
                        if (event.getAffectedIdentifiers() != null) {
                            log.log(Level.INFO,
                                    "Received aggregation request for {0} between {1} and {2}",
                                    new Object[]{
                                        Arrays.toString(event.getAffectedIdentifiers().toArray()), event.getStartDate(), event.getEndDate()
                                    }
                            );

                            for (Integer companyId : event.getAffectedIdentifiers()) {
                                try {
                                    // Code which handles the aggregation between start and end, does work in parts, segmented by REQUIRES_NEW when saving new data
                                    handler.handle(companyId, event.getStartDate(), event.getEndDate(), aggregationController);
                                } catch (Exception ex) {
                                    log.log(Level.WARNING, "Error while handling aggregation for company(id={0}) - Error: {1}", new Object[]{companyId, ex.getMessage()});
                                }
                            }

                            log.log(Level.INFO,
                                    "Finished aggregation request for {0} between {1} and {2}",
                                    new Object[]{
                                        Arrays.toString(event.getAffectedIdentifiers().toArray()), event.getStartDate(), event.getEndDate()
                                    }
                            );
                        }
                    }
                }

            } catch (Exception ex) {
                log.log(Level.SEVERE, "onMessage encountered an Exception", ex);
            }
        }
    }

}
```

### LogDump
```
[2015-07-28T11:25:15.037+0200] [Payara 4.1] [SEVERE] [] [javax.resourceadapter.mqjmsra.outbound.connection] [tid: _ThreadID=283 _ThreadName=p: thread-pool-1; w: 8] startTransaction (XA) on JMSService:jmsdirect failed for connectionId:2191087890899026944 due to Unknown JMSService server error NOT_FOUND: com.sun.messaging.jmq.jmsserver.util.BrokerException: Unknown XID 4C696576656E506F656C6D616E2C7365727665722C50333730302C00CA0000006E54B4D34C696576656E506F656C6D616E2C7365727665722C5033373030

[2015-07-28T11:25:15.039+0200] [Payara 4.1] [WARNING] [jts.resource_outside_transaction] [javax.enterprise.system.core.transaction.com.sun.jts.jta] [tid: _ThreadID=283 _ThreadName=p: thread-pool-1; w: 8] [[
  JTS5041: The resource manager is doing work outside a global transaction
javax.transaction.xa.XAException
	at com.sun.messaging.jms.ra.DirectXAResource.sendStartToBroker(DirectXAResource.java:879)
	at com.sun.messaging.jms.ra.DirectXAResource.start(DirectXAResource.java:835)
	at com.sun.jts.jta.TransactionState.startAssociation(TransactionState.java:364)
	at com.sun.jts.jta.TransactionImpl.enlistResource(TransactionImpl.java:212)
	at com.sun.enterprise.transaction.JavaEETransactionImpl.enlistResource(JavaEETransactionImpl.java:660)
	at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.enlistXAResource(JavaEETransactionManagerSimplified.java:1313)
	at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.enlistResource(JavaEETransactionManagerSimplified.java:371)
	at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.enlistComponentResources(JavaEETransactionManagerSimplified.java:1339)
	at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.enlistComponentResources(JavaEETransactionManagerSimplified.java:501)
	at com.sun.ejb.containers.EJBContainerTransactionManager.useClientTx(EJBContainerTransactionManager.java:373)
	at com.sun.ejb.containers.EJBContainerTransactionManager.preInvokeTx(EJBContainerTransactionManager.java:253)
	at com.sun.ejb.containers.BaseContainer.preInvokeTx(BaseContainer.java:4524)
	at org.glassfish.ejb.mdb.MessageBeanContainer.beforeMessageDelivery(MessageBeanContainer.java:1164)
	at org.glassfish.ejb.mdb.MessageBeanListenerImpl.beforeMessageDelivery(MessageBeanListenerImpl.java:77)
	at com.sun.enterprise.connectors.inbound.MessageEndpointInvocationHandler.invoke(MessageEndpointInvocationHandler.java:139)
	at com.sun.proxy.$Proxy828.beforeDelivery(Unknown Source)
	at com.sun.messaging.jms.ra.OnMessageRunner.run(OnMessageRunner.java:262)
	at com.sun.enterprise.connectors.work.OneWork.doWork(OneWork.java:107)
	at com.sun.corba.ee.impl.threadpool.ThreadPoolImpl$WorkerThread.performWork(ThreadPoolImpl.java:497)
	at com.sun.corba.ee.impl.threadpool.ThreadPoolImpl$WorkerThread.run(ThreadPoolImpl.java:540)
Caused by: com.sun.messaging.jmq.jmsservice.JMSServiceException: startTransaction: start transaction failed. Connection ID: 2191087890899026944, session ID: 0, XID: (Available at FINE log level) Caused by:com.sun.messaging.jmq.jmsserver.util.BrokerException: Unknown XID 4C696576656E506F656C6D616E2C7365727665722C50333730302C00CA0000006E54B4D34C696576656E506F656C6D616E2C7365727665722C5033373030
	at com.sun.messaging.jmq.jmsserver.service.imq.JMSServiceImpl.startTransaction(JMSServiceImpl.java:1426)
	at com.sun.messaging.jms.ra.DirectXAResource.sendStartToBroker(DirectXAResource.java:854)
	... 19 more
Caused by: com.sun.messaging.jmq.jmsserver.util.BrokerException: Unknown XID 4C696576656E506F656C6D616E2C7365727665722C50333730302C00CA0000006E54B4D34C696576656E506F656C6D616E2C7365727665722C5033373030
	at com.sun.messaging.jmq.jmsserver.data.protocol.ProtocolImpl.startTransaction(ProtocolImpl.java:714)
	at com.sun.messaging.jmq.jmsserver.service.imq.JMSServiceImpl.startTransaction(JMSServiceImpl.java:1415)
	... 20 more
]]
```

**Comment by dorpsidioot** [comment]
Extra information: this seems to only happen when JMS messages are still in the queue during a redeploy. I'm trying to reproduce this behaviour.

**Comment by smillidge** [comment]
Hi is the exception above all the exception as I don't see any of your code in there?

**Comment by smillidge** [comment]
Try precreating the producer in a @PostConstruct method as you are creating multiple producers in a single transaction which is probably not helpful.

**Comment by dorpsidioot** [comment]
The exception is located in the log dump part of the message. THe issue here is that I can't catch this exception and analyse it, since it happens before the message gets delivered it seems. And the weird part is that I don't seem to lose any messages, they get delivered anyways. Though I regularly get this issue when messages get redelivered after redeploying the application. I will change the code to using a single producer instead of recreating it constantly.

**Comment by smillidge** [comment]
Could you add some of the log messages either side so we can get some context to see when this is happening.

**Comment by dorpsidioot** [comment]
I will try and create a test application which injects messages with a unique ID in the queue in order to make sure all were received 1 on 1, and try to force this issue to appear.

**Comment by dorpsidioot** [comment]
Can be closed, since 4.1.153 I have no longer encountered this issue.

**Comment by smillidge** [comment]
Thanks for letting us know. Glad to hear it is resolved.
